Packs/Symantec_Advanced_Threat_Protection/Integrations/integration-Symantec_Advanced_Threat_Protection.yml

```diff
@@ -261,14 +275,117 @@ script:
+        } else if (unit.indexOf('hour') >= 0) {
+            d.setTime(d.getTime() - (number * 60 * 60 * 1000));
+        } else if (unit.indexOf('day') >= 0) {
+            d.setTime(d.getTime() - (number * 24 * 60 * 60 * 1000));
+        } else if (unit.indexOf('month') >= 0) {
+            d.setTime(d.getTime() - (number * 30 * 24 * 60 * 60 * 1000));
+        } else if (unit.indexOf('month') >= 0) {
+            d.setTime(d.getTime() - (number * 365 * 30 * 24 * 60 * 60 * 1000));
+        }
+        return d;
+    };
 
-    // The command input arg holds the command sent from the user.
+    var getIncidentsFromIncidentsResult = function(res) {
+        return res.map(function(r){
+            return {
+                name: r.summary,
+                rawJSON: JSON.stringify(r)
+            };
+        });
+    };
 
+    var getIncidentsFromEventsResult = function(res) {
+        return res.map(function(r){
+            var name;
+            if ('message' in r){
+                name = r.message;
+            } else if ('uuid' in r){
+                name = r.uuid;
+            } else {
+                name = r.type_id;
+            }
+            return {
+                name: name,
+                rawJSON: JSON.stringify(r)
+            };
+        });
+    };
+
+    var fetchIncidents = function(type, query, limit, first_fetch_time) {
+
+        if (type != 'incidents' && type != 'incidentevents' && type != 'events'){
+            throw 'Unknown incident type.';
+        }
+        var DAY_IN_MILLIS = 24*60*60*1000;
+
+        var lastRun = getLastRun();
+        var now = new Date();
+        var endTime = now.toISOString();
+        var startTime;
+
+        if (lastRun && lastRun.time && lastRun.time !== '') {
+            startTime = lastRun.time;
+        } else if (first_fetch_time) {
+            startTime = parseOffsetDate(first_fetch_time).toISOString();
+        } else {
+            var n = now;
+            n.setTime(now.getTime() - 1 * DAY_IN_MILLIS);
+            startTime = n.toISOString();
+        }
+        var next = false;
+        var p = {verb: 'query'};
+        add(p, 'where', query);
+        addTime(p, 'start_time', startTime);
+        addTime(p, 'end_time', endTime);
+        if (limit) {
+            limit = parseInt(limit);
+            p.limit = Math.min(limit, 100);
+        }
+        var incidents = [];
+        do {
+            add(p, 'next', next);
+            var res = doReq('POST', type, p);
+            if (res.obj.result && res.obj.result.length !== 0) {
+                if (type == 'incidents') {
+                    incidents = incidents.concat(getIncidentsFromIncidentsResult(res.obj.result));
+                } else {
+                    incidents = incidents.concat(getIncidentsFromEventsResult(res.obj.result));
+                }
+                next = 'next' in res.obj ? res.obj.next : null;
+            }
+        } while (next && limit && incidents.length < limit);
```

Comment:
Not sure I understand the logic of the `do.. while` loop. In what scenario will we do the loop more than once? If there is a limit set then the number of results we will get back from the API call will be less or equals to the limit.
If it's less, then no `next` will return. If it's equal, then `next` will be returned but `incidents.length == limit` won't it?
What I suggest is keep the `next` parameter in the `setLastRun` and in the next fetch use it to fetch the next batch of incidents, do this until there's no `next` parameter then proceed as usual. We're using this logic in a few integrations like [Splunkpy](https://github.com/demisto/content/blob/e1309967af65fdca92ee0035487030092acbb8d4/Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py#L505) if you want to have a look.

I'm happily available on the DFIR channel for any questions about this.